README
Module to replace Math::BigInt's core math routines by using Math::GMP.

(C) ..--- ----- ----- .---- by - . .-.. ... .-.-.-  All rights reserved.

This package contains a replacement (drop-in) module for Math::BigInt's core,
Math::BigInt::Calc.pm. It needs the new versions of Math::BigInt and
Math::BigFloat as they are from Perl 5.7.x onwards.

How to install (replace the version below with the current one):

	tar -xzf Math-BigInt-GMP-1.10.tar.gz
	cpansign --verify

If the signature does not verify ok, you might to install Module::Signature
and get my key from <http://bloodgate.com/tels.asc>.

If the signature still does not verify okay, notify me. DO NO INSTALL this
package when the signature is BAD - your system might get compromised!

If the signature is good, do:

	perl Makefile.PL
	make
	make test

and then as root:

	make install


To use this module, replace:

	use Math::BigInt;

with this:

	use Math::BigInt lib => 'GMP';

or alternatively something like:
	
	use Math::BigInt lib => 'GMP,Pari,BitVect';

Please read the documentation under PERFORMANCE in Math::BigInt on when to
use this module and when not. As of v1.10 and Math::BigInt v1.64/v1.64, using
"lib => 'GMP'" will actually reduce the memory consumption of your program,
and thanx to the fact that we no longer use Math::GMP but have our own XS
code, almost all operations should be faster than just using Calc (the
default pure-perl library used in Math::BigInt).

So, the following:
	
	use Math::BigInt lib => 'GMP';

can be heartily recommended.

More information including HTMLified help and preliminary benchmark results
are available under:

	http://bloodgate.com/perl/bigint/

Please send me test-reports, your experiences with this and your ideas - I love
to hear about my work!

Tels <http://bloodgate.com/>


CHANGES
2010-09-23 v1.32 rafl 5559 tests
  * Re-upload 1.31 as a stable release without further changes.

2010-09-21 v1.31 rafl 5559 tests  DEVELOPMENT RELEASE
  * Add hooks for Storable (de-)serialisation.
  * Avoid failure in the destructor if someone blessed nonsense into our class.

2010-09-20 v1.30 rafl 5558 tests
  * Re-upload 1.29 as a stable release without further changes.

2010-09-19 v1.29 rafl 5558 tests  DEVELOPMENT RELEASE
  * Attempt to fix a bug in the bundled version Devel::CheckLib.
    It used to ignore @Config{qw(ccflags ldflags)} and only tried to look for
    headers and libraries with the compiler's default include- and lib-paths as
    well as those explicitly asked for by the user.

2010-09-17 v1.28 rafl 5558 tests
  * Re-upload 1.27 without further changes as a stable release.

2010-09-15 v1.27 rafl 5558 tests  DEVELOPMENT RELEASE
  * Try to support perls older than 5.8.8 again.
    Tested with 5.8.7 and 5.6.2.

2010-09-14 v1.26 rafl 5558 tests  DEVELOPMENT RELEASE
  * Error out early if libgmp or gmp.h are missing.
  * Clone Math::BigInt::GMP instances on thread cloning.
    This should make the module threadsafe.

2010-09-10 v1.25 rafl 5536 tests
  * Fix tests with Math::BigInt >= 1.90 and depend on it.

2007-07-31 v1.24 Tels 5530 tests
  * apply patch for warnings about ptr size mismatch under Cygwin (thanx Reini Urban!)
  * make it work under 5.6.x again by defining SvUOK() (Thanx Marcus Holland-Moritz
    and Reini Urban!)

2007-07-25 v1.23 Tels 5527 tests
  * require Math::BigInt 1.87
  * fix for _new() (appeared under Cygwin, but possible others, thanx
    Linda W. (report) and Reini Urban (patch)!)

2007-06-01 v1.22 Tels 5527 tests
  * require Math::BigInt 1.86
  * support api_version() 2 by adding _nok()
  * fix compilation issues on Mac/Darwin
  * _log_int() modifies it's argument instead of just returning a
    different object as result
  * speed up _log_int() greatly by taking a guess of the result
    and then improve it, instead of startig with 1 and going up. This means
    it takes now a more or less constant time, instead of a time proportional
    to the size/value of the result:

 Using Math::BigInt::GMP v1.21
  baselen 2:     3s (3.22 usr +  0.00 sys =  3.22 CPU) @ 7270/s (n=23411)
  baselen 2 big: 3s (3.11 usr +  0.06 sys =  3.17 CPU) @  962/s (n=3051)
  baselen 3:     3s (3.20 usr +  0.00 sys =  3.20 CPU) @ 1304/s (n=4173)
  baselen 7:     3s (3.20 usr +  0.00 sys =  3.20 CPU) @ 3306/s (n=10582)
  baselen 8:     3s (3.14 usr +  0.00 sys =  3.14 CPU) @ 3769/s (n=11836)
  baselen 11:    3s (3.24 usr +  0.00 sys =  3.24 CPU) @ 4750/s (n=15392)
  baselen 14:    3s (3.20 usr +  0.00 sys =  3.20 CPU) @ 5835/s (n=18673)
  baselen 20:    3s (3.10 usr +  0.03 sys =  3.13 CPU) @ 7621/s (n=23855)
  baselen 3 big: 4s (3.17 usr +  0.00 sys =  3.17 CPU) @  320/s (n=1016)

 Using Math::BigInt::GMP v1.22
  baselen 2:     3s (3.15 usr +  0.00 sys =  3.15 CPU) @ 16290/s (n=51316)
  baselen 2 big: 4s (3.21 usr +  0.01 sys =  3.22 CPU) @ 15933/s (n=51306)
  baselen 3:     4s (3.12 usr +  0.02 sys =  3.14 CPU) @ 15555/s (n=48844)
  baselen 7:     4s (3.15 usr +  0.07 sys =  3.22 CPU) @ 15658/s (n=50420)
  baselen 8:     3s (3.18 usr +  0.01 sys =  3.19 CPU) @ 15610/s (n=49797)
  baselen 11:    3s (3.14 usr +  0.00 sys =  3.14 CPU) @ 15555/s (n=48844)
  baselen 14:    3s (3.15 usr +  0.00 sys =  3.15 CPU) @ 15506/s (n=48844)
  baselen 20:    3s (3.14 usr +  0.01 sys =  3.15 CPU) @ 15506/s (n=48844)
  baselen 3 big: 3s (3.10 usr +  0.04 sys =  3.14 CPU) @ 15555/s (n=48844)

2007-04-17 v1.21 Tels 5488 tests
  * add _as_oct(), _from_oct(), _alen(), _1ex() and some _root() tests
  * require Math::BigInt 1.83
  * support api_version() by adding _1ex() and _alen()
  * _new(): take a shortcut if the passed an IV (integer value)

2007-04-09 v1.20 Tels 5351 tests
  * remove PREREQ_FATAL because the toolchain is broken and cannot handle it
  * take over tests from MBI 1.82 and require it
  * require Perl 5.6.2 as minimum
  * speed up _zeros():
    + use Newx() instead of a full-blown SV for temp storage
    + no need to allocate temp storage for numbers < 10
  * put _len() into XS code, making $x->length() faster (about 30% for "123",
    less for longer numbers as the binary=>decimal conversion dominates)
  * add POD tests
  * add MANIFEST.SKIP

2007-01-27 v1.19 Tels 5339 tests
  * add support for octal
  * take over tests from MBI 1.78 and require it

Older Changelog:

2001-07-22 v1.00 Tels
 * First version (basically working with some quirks)
2001-08-06 v1.01 Tels
 * first release
 * fixed all the bugs in v1.00
 * taken over tests from BigInt v1.40
2001-09-02 v1.02 Tels
 * removed auto-export and added empty import()
 * taken over tests from BigInt v1.42
2001-11-01 v1.03 Tels
 * taken over tests from BigInt v1.45 
 * added _mod() for more speed for $x % $y

	#!/usr/bin/perl -w
	use lib 'lib';
	#use lib '../Math-BigInt-GMP-1.02/lib';
	use Math::BigInt lib => 'GMP';
	use Benchmark;
	my $digits = 1024;
	my $x = Math::BigInt->new('1' . '0' x $digits);
	my $y = Math::BigInt->new('3' . '0' x ($digits - 2) . '3');
	my $u = Math::BigInt->new('3');
	timethese ( 2000,
	  {
	  mod_l => sub { $z = $x % $y, },
	  mod_s => sub { $z = $x % $u, },
	  div_l => sub { ($z,$r) = $x->copy()->bdiv($y), },
	  } );

	On a 1 Ghz Athlon with v1.45 of BigInt in ops/s:
		v1.02	v1.03
	mod_s    1100	 2350 
	mod_l    1111    2325
	div_l    1260    1300

2001-11-01 v1.04a Tels (never released)
  * _is_odd()/_is_even() use $two instead of 2: 5600 op/s instead of 4700
2002-01-26 v1.04 Tels
  * use $zero,$one,$two etc instead of 0,1,2 in some routines
  * tests from Math::BigInt v1.50
  * bypass Math::GMP's overload interface and use Math::GMP::gmp_foo() directly
  * added _gcd() and _fac() for more speed in bgcd() and bfac(), respectively
2002-02-16 v1.05 Tels
  * tests from Math::BigInt v1.51
  * replaced _core_lib() by config()->{lib}
  * added _and, _or, _xor (using Math::GMP internal methods)

lib/Math/BigInt/GMP.pm
###############################################################################
# core math lib for BigInt, representing big numbers by the GMP library

package Math::BigInt::GMP;

use strict;
use 5.006002;
# use warnings; # dont use warnings for older Perls

use vars qw/$VERSION/;

$VERSION = '1.32';

use XSLoader;
XSLoader::load "Math::BigInt::GMP", $VERSION;

sub import { }			# catch and throw away
sub api_version() { 2; }	# we are compatible with MBI v1.83 and up

BEGIN
  {
  # both _num and _str just return a string
  *_str = \&_num;
  }

##############################################################################
# Routines not present here are in GMP.xs

sub _digit
  {
  # return the nth digit, negative values count backward; this is costly!
  my ($c,$x,$n) = @_;

  $n++; substr( Math::BigInt::GMP::_num($c,$x), -$n, 1 );
  }

###############################################################################
# routine to test internal state for corruptions

sub _check
  {
  # no checks yet, pull it out from the test suite
  my ($x) = $_[1];
  return "$x is not a reference to Math::BigInt::GMP"
   if ref($x) ne 'Math::BigInt::GMP';
  0;
  }

sub _log_int
  {
  my ($c,$x,$base) = @_;

  # X == 0 => NaN
  return if _is_zero($c,$x);

  $base = _new($c,2) unless defined $base;
  $base = _new($c,$base) unless ref $base;

  # BASE 0 or 1 => NaN
  return if _is_zero($c,$base) || _is_one($c,$base);

  my $cmp = _acmp($c,$x,$base); 	# X == BASE => 1
  if ($cmp == 0)
    {
    # return one
    return (_one($c), 1);
    }
  # X < BASE
  if ($cmp < 0)
    {
    return (_zero($c),undef);
    }

  # Compute a guess for the result based on:
  # $guess = int ( length_in_base_10(X) / ( log(base) / log(10) ) )
  my $len = _alen($c,$x);
  my $log = log( _num($c,$base) ) / log(10);

  # calculate now a guess based on the values obtained above:
  my $x_org = _copy($c,$x);

  # keep the reference to $x, modifying it in place
  _set($c, $x, int($len / $log) - 1);

  my $trial = _pow ($c, _copy($c, $base), $x);
  my $a = _acmp($c,$trial,$x_org);

  if ($a == 0)
    {
    return ($x,1);
    }
  elsif ($a > 0)
    {
    # too big, shouldn't happen
    _div($c,$trial,$base); _dec($c, $x);
    }

  # find the real result by going forward:
  my $base_mul = _mul($c, _copy($c,$base), $base);
  my $two = _two($c);

  while (($a = _acmp($c, $trial, $x_org)) < 0)
    {
    _mul($c,$trial,$base_mul); _add($c, $x, $two);
    }

  my $exact = 1;
  if ($a > 0)
    {
    # overstepped the result
    _dec($c, $x);
    _div($c,$trial,$base);
    $a = _acmp($c,$trial,$x_org);
    if ($a > 0)
      {
      _dec($c, $x);
      }
    $exact = 0 if $a != 0;
    }

  ($x,$exact);
  }

sub STORABLE_freeze {
    my ($self, $cloning) = @_;
    return Math::BigInt::GMP->_num($self);
}

sub STORABLE_thaw {
    my ($self, $cloning, $serialized) = @_;
    Math::BigInt::GMP->_new_attach($self, $serialized);
    return $self;
}

1;
__END__

=pod

=head1 NAME

Math::BigInt::GMP - Use the GMP library for Math::BigInt routines

=head1 SYNOPSIS

Provides support for big integer calculations via means of the GMP c-library.

Math::BigInt::GMP now no longer uses Math::GMP, but provides it's own XS layer
to access the GMP c-library. This cut's out another (perl sub routine) layer
and also reduces the memory footprint by not loading Math::GMP and Carp at
all.

=head1 LICENSE
 
This program is free software; you may redistribute it and/or modify it under
the same terms as Perl itself. 

=head1 AUTHOR

Tels <http://bloodgate.com/> in 2001-2007.

Thanx to Chip Turner for providing Math::GMP, which was inspiring my work.

=head1 SEE ALSO

L<Math::BigInt>, L<Math::BigInt::Calc>.

=cut


Makefile.PL
use ExtUtils::MakeMaker;
use lib 'inc'; # load our bundled version of Devel::CheckLib
use Devel::CheckLib;

#printf "Your OS is [%s]!\n---\n", $Config::Config{'osname'};  ### for testing purpose only

if (($Config::Config{'osname'} eq 'darwin') &! (join (' ', @ARGV) =~ /LDDLFLAGS/))
  {
  ### reads pre-existing LDDLFLAGS for a darwin system as defined in
  ###  /System/Library/Perl/x.x.x/darwin-thread-multi-2level/Config.pm
  ###  where x.x.x is the perl version, which is 5.8.1 for MacOS X 10.3,
  ###  and appends ' -read_only_relocs suppress' 
  ### default behavior if LDDLFLAGS is defined
  push @ARGV, sprintf "LDDLFLAGS=%s -read_only_relocs suppress", $Config::Config{'lddlflags'};
  }

sub MY::metafile {
  package MY;
  my $self = shift;
  if (!eval { require YAML; 1 })
    {
    warn ("YAML not installed, make dist will not override metafile");
    return $self->SUPER::metafile_target(@_);
    }

  require YAML::Node;
  my $node = new YAML::Node {};

  $node->{name}                 = $self->{DISTNAME};
  $node->{author}               = 'Tels';
  $node->{version}              = $self->{VERSION};
  $node->{version_from}         = $self->{VERSION_FROM};
  $node->{license}              = 'perl';
  $node->{distribution_type}    = 'module';
  $node->{generated_by}         = "$self->{DISTNAME} version $self->{VERSION}";
  $node->{installdirs}          = 'site';
  $node->{requires}             = $self->{PREREQ_PM};

  my $dump = YAML::Dump( $node );

  $dump =~ s/^(.*)$/\t\$(NOECHO) \$(ECHO) "$1" >>META.yml/gm;
  $dump =~ s/>>META\.yml/>META.yml/;

  "metafile:\n$dump";
  }

check_lib_or_exit(lib => 'gmp', header => 'gmp.h', @ARGV);

WriteMakefile(
    'NAME'		=> 'Math::BigInt::GMP',
    'VERSION_FROM'	=> 'lib/Math/BigInt/GMP.pm',
    'PREREQ_PM'		=> {
				Math::BigInt => 1.90,
				XSLoader => 0.02,
			   },
    'LIBS'		=> ['-lgmp'],
    'SIGN' => 1,
);
